sdks/java/harness/src/main/java/org/apache/beam/fn/harness/logging/BeamFnLoggingClient.java

```diff
@@ -117,75 +188,216 @@ public BeamFnLoggingClient(
+            }
+            continue;
+          }
+
+          // Batch together as many log messages as possible that are held within the buffer
+          BeamFnApi.LogEntry.List.Builder builder =
+              BeamFnApi.LogEntry.List.newBuilder().addLogEntries(logEntry);
+          bufferedLogEntries.drainTo(additionalLogEntries);
+          builder.addAllLogEntries(additionalLogEntries);
+          outboundObserver.onNext(builder.build());
+          additionalLogEntries.clear();
+        }
+        if (inboundObserverCompletion.isDone()) {
+          try {
+            // If the inbound observer failed with an exception, get() will throw an
+            // ExecutionException.
+            inboundObserverCompletion.get();
+            // Otherwise it is an error for the server to close the stream before we closed our end.
+            throw new IllegalStateException(
+                "Logging stream terminated unexpectedly with success before it was closed by the client.");
+          } catch (ExecutionException e) {
+            throw new IllegalStateException(
+                "Logging stream terminated unexpectedly before it was closed by the client with error: "
+                    + e.getCause());
+          } catch (InterruptedException e) {
+            // Should never happen because of the isDone check.
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+          }
+        }
+      } catch (Throwable t) {
+        thrown = t;
+        throw new RuntimeException(t);
+      } finally {
+        if (thrown == null) {
+          outboundObserver.onCompleted();
+        } else {
+          outboundObserver.onError(thrown);
+        }
+        channel.shutdown();
+        boolean shutdownFinished = false;
+        try {
+          shutdownFinished = channel.awaitTermination(10, SECONDS);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        } finally {
+          if (!shutdownFinished) {
+            channel.shutdownNow();
+          }
+        }
+      }
+    }
+
+    public void softClose() {
+      softClosing.complete(COMPLETED);
+    }
+
+    public void hardClose() {
+      streamPhaser.forceTermination();
+    }
+
+    private class LogControlObserver
+        implements ClientResponseObserver<BeamFnApi.LogEntry, BeamFnApi.LogControl> {
+
+      @Override
+      public void beforeStart(ClientCallStreamObserver<BeamFnApi.LogEntry> requestStream) {
+        requestStream.setOnReadyHandler(streamPhaser::arrive);
+      }
+
+      @Override
+      public void onNext(BeamFnApi.LogControl value) {}
+
+      @Override
+      public void onError(Throwable t) {
+        inboundObserverCompletion.completeExceptionally(t);
+        hardClose();
+      }
+
+      @Override
+      public void onCompleted() {
+        inboundObserverCompletion.complete(COMPLETED);
+        hardClose();
+      }
+    }
   }
 
   @Override
+  @SuppressWarnings("nullness")
```

Comment:
there is `Preconditions.checkNotNull` below, is this SuppressWarnings still needed (same as `terminationFuture`  below)

Reply:
For some reason with the Preconditions.checkNotNull I still got errors. Interestingly I tried casting explicitly to @NonNull afterwards and it said both that it wasn't safe and that it was unnecessary :P
I tried going to CompletableFuture<Object> instead of CompletableFuture<?> in case that was messing it up but it didn't help.
I was able to get it to work by removing the Nullable by initializing it in the constructor instead of via the helper start method.  To do so I had to add some additional underinitialization annotations.
PTAL